### ISSUE TITLE
Can't call method on `impl Trait` without explicitly passing generic arguments

### ISSUE THREAD

**Issue by rodrimati1992** [rust-lang/rust#66057]
The [**next comment**](#issuecomment-602129468)  has a minimal example(the one in this comment is a bit complex).

### Old example

This code:

https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=d001a5aeff1523dcec587c3e6e7ea81d

```rust
trait Hello<Name> {
    fn hello(&self, _: Name){}
}

struct A;
struct B;
struct C;

impl<T0, T1> Hello<A> for (T0, T1) {}
impl<T0, T1> Hello<B> for (T0, T1) {}
impl<T0, T1> Hello<C> for (T0, T1) {}

trait Tuple2:Hello<A> + Hello<B> + Hello<C>{}

impl<This> Tuple2 for This 
where 
    This:Hello<A> + Hello<B> + Hello<C>
{}


fn returns_impl() -> impl Tuple2 {
    (0, 10)
}


fn main() {
    {
        let tup = returns_impl();
        tup.hello(A); // This is the first one that errors
        tup.hello(B); // This is the second one that errors
        tup.hello(C);
        Hello::<A>::hello(&tup,A);
        Hello::<B>::hello(&tup,B);
        Hello::<C>::hello(&tup,C);
    }
    {// It all works fine with dynamic dispatch
        let tup:&dyn Tuple2= &returns_impl();
        tup.hello(A);
        tup.hello(B);
        tup.hello(C);
        Hello::<A>::hello(tup,A);
        Hello::<B>::hello(tup,B);
        Hello::<C>::hello(tup,C);
    }
}

```
Causes this error:
```
   Compiling playground v0.0.1 (/playground)
error[E0308]: mismatched types
  --> src/main.rs:29:19
   |
29 |         tup.hello(A); // This is the first one that errors
   |                   ^ expected struct `C`, found struct `A`
   |
   = note: expected type `C`
              found type `A`

error[E0308]: mismatched types
  --> src/main.rs:30:19
   |
30 |         tup.hello(B); // This is the second one that errors
   |                   ^ expected struct `C`, found struct `B`
   |
   = note: expected type `C`
              found type `B`

error: aborting due to 2 previous errors
```
On both the stable (1.38) and nightly (1.40.0 2019-11-02) versions in the playground.

I expected this code to compile without errors.

**Comment by rodrimati1992** [comment]
Here is a minimized version of this error:
https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=88c617f0e3c7a5edc0734dfe9c8e3e6c
```rust
fn make()-> impl Extend<&'static str> + Extend<char> {
    String::new()
}
fn main(){
    make().extend(std::iter::repeat("string").take(10));
    make().extend(std::iter::repeat('c').take(10));
}
```
```
error[E0271]: type mismatch resolving `<std::iter::Take<std::iter::Repeat<&str>> as std::iter::IntoIterator>::Item == char`
 --> src/main.rs:5:12
  |
5 |     make().extend(std::iter::repeat("string").take(10));
  |            ^^^^^^ expected `&str`, found `char`
```

**Comment by benjaminp** [comment]
This was fixed by #73905.

**Comment by jyn514** [comment]
Both the original and reduced example now compile.
